File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/VoucherSelectionBottomSheet.java
package com.cinema.ticket_booking.ui.booking;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.cinema.ticket_booking.R;
import com.cinema.ticket_booking.data.model.response.VoucherSummary;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class VoucherSelectionBottomSheet extends BottomSheetDialogFragment {

    public interface OnVoucherSelectedListener {
        void onVoucherSelected(VoucherSummary voucher);
        void onManualVoucherApplied(String code);
    }

    private List<VoucherSummary> allVouchers = new ArrayList<>();
    private double cartTotal;
    private OnVoucherSelectedListener listener;

    private RecyclerView rvVouchers;
    private EditText etManualCode;
    private Button btnApplyManual;

    public static VoucherSelectionBottomSheet newInstance(List<VoucherSummary> vouchers, double cartTotal) {
        VoucherSelectionBottomSheet fragment = new VoucherSelectionBottomSheet();
        fragment.allVouchers = vouchers;
        fragment.cartTotal = cartTotal;
        return fragment;
    }

    public void setListener(OnVoucherSelectedListener listener) {
        this.listener = listener;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_voucher_bottom_sheet, container, false);
        rvVouchers = view.findViewById(R.id.rvVouchers);
        etManualCode = view.findViewById(R.id.etManualCode);
        btnApplyManual = view.findViewById(R.id.btnApplyManual);

        rvVouchers.setLayoutManager(new LinearLayoutManager(requireContext()));

        btnApplyManual.setOnClickListener(v -> {
            String code = etManualCode.getText().toString().trim();
            if(!code.isEmpty() && listener != null) {
                listener.onManualVoucherApplied(code.toUpperCase());
                dismiss();
            } else {
                Toast.makeText(getContext(), "Vui lòng nhập mã hợp lệ", Toast.LENGTH_SHORT).show();
            }
        });

        setupAdapter();

        return view;
    }

    private void setupAdapter() {
        List<VoucherItem> available = new ArrayList<>();
        List<VoucherItem> disabled = new ArrayList<>();

        for (VoucherSummary v : allVouchers) {
            if (!"AVAILABLE".equals(v.getStatus())) {
                disabled.add(new VoucherItem(v, 0, "Chỉ áp dụng với mã khả dụng / Còn hạn."));
                continue;
            }
            if (v.getMinOrder() > 0 && cartTotal < v.getMinOrder()) {
                disabled.add(new VoucherItem(v, 0, "Đơn tối thiểu " + String.format("%,.0f₫", v.getMinOrder())));
                continue;
            }

            double actualDiscount = 0;
            if ("PERCENTAGE".equals(v.getDiscountType())) {
                actualDiscount = (cartTotal * v.getDiscountValue()) / 100.0;
            } else {
                actualDiscount = v.getDiscountValue();
            }
            if (v.getMaxDiscount() > 0 && actualDiscount > v.getMaxDiscount()) {
                actualDiscount = v.getMaxDiscount();
            }
            available.add(new VoucherItem(v, actualDiscount, null));
        }

        Collections.sort(available, (a, b) -> Double.compare(b.actualDiscount, a.actualDiscount));

        List<VoucherItem> combined = new ArrayList<>();
        combined.addAll(available);
        combined.addAll(disabled);

        rvVouchers.setAdapter(new SheetAdapter(combined));
    }

    private class VoucherItem {
        VoucherSummary summary;
        double actualDiscount;
        String disabledReason;

        VoucherItem(VoucherSummary summary, double actualDiscount, String disabledReason) {
            this.summary = summary;
            this.actualDiscount = actualDiscount;
            this.disabledReason = disabledReason;
        }
    }

    private class SheetAdapter extends RecyclerView.Adapter<SheetAdapter.VH> {
        private List<VoucherItem> items;

        SheetAdapter(List<VoucherItem> items) {
            this.items = items;
        }

        @NonNull
        @Override
        public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_voucher, parent, false);
            return new VH(v);
        }

        @Override
        public void onBindViewHolder(@NonNull VH holder, int position) {
            VoucherItem item = items.get(position);
            VoucherSummary v = item.summary;

            holder.tvCode.setText(v.getCode());
            holder.tvDescription.setText(v.getDescription());
            holder.tvValidTo.setText("HSD: " + (v.getEndDate() != null ? v.getEndDate().substring(0, 10) : "N/A"));

            if (item.disabledReason == null) {
                // Khả dụng
                holder.itemView.setAlpha(1.0f);
                holder.tvStatus.setText("Khả dụng");
                holder.tvStatus.setBackgroundTintList(android.content.res.ColorStateList.valueOf(0x334CAF50));
                holder.tvStatus.setTextColor(0xFF4CAF50);
                holder.tvDiscount.setText(String.format("-%,.0f₫", item.actualDiscount));
                
                holder.itemView.setOnClickListener(view -> {
                    if (listener != null) {
                        listener.onVoucherSelected(v);
                        dismiss();
                    }
                });
            } else {
                // Disabled
                holder.itemView.setAlpha(0.5f);
                holder.tvStatus.setText(item.disabledReason);
                holder.tvStatus.setBackgroundTintList(android.content.res.ColorStateList.valueOf(0x33888888));
                holder.tvStatus.setTextColor(0xFF888888);
                
                String discount;
                if ("PERCENTAGE".equalsIgnoreCase(v.getDiscountType())) {
                    discount = "-" + (int) v.getDiscountValue() + "%";
                } else {
                    discount = String.format("-%,.0f₫", v.getDiscountValue());
                }
                holder.tvDiscount.setText(discount);
                holder.itemView.setOnClickListener(null);
            }
        }

        @Override
        public int getItemCount() {
            return items.size();
        }

        class VH extends RecyclerView.ViewHolder {
            TextView tvCode, tvDescription, tvValidTo, tvDiscount, tvStatus;
            VH(@NonNull View itemView) {
                super(itemView);
                tvCode = itemView.findViewById(R.id.tvCode);
                tvDescription = itemView.findViewById(R.id.tvDescription);
                tvValidTo = itemView.findViewById(R.id.tvValidTo);
                tvDiscount = itemView.findViewById(R.id.tvDiscount);
                tvStatus = itemView.findViewById(R.id.tvStatus);
            }
        }
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/VoucherSelectionBottomSheet.kt
package com.cinema.ticket_booking.ui.booking

import android.content.res.ColorStateList
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.cinema.ticket_booking.R
import com.cinema.ticket_booking.data.model.response.VoucherSummary
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import java.util.ArrayList
import java.util.Collections
import java.util.Locale

class VoucherSelectionBottomSheet : BottomSheetDialogFragment() {

    interface OnVoucherSelectedListener {
        fun onVoucherSelected(voucher: VoucherSummary)
        fun onManualVoucherApplied(code: String)
    }

    private var allVouchers: List<VoucherSummary> = ArrayList()
    private var cartTotal: Double = 0.0
    private var listener: OnVoucherSelectedListener? = null

    private lateinit var rvVouchers: RecyclerView
    private lateinit var etManualCode: EditText
    private lateinit var btnApplyManual: Button

    companion object {
        fun newInstance(vouchers: List<VoucherSummary>, cartTotal: Double): VoucherSelectionBottomSheet {
            val fragment = VoucherSelectionBottomSheet()
            fragment.allVouchers = vouchers
            fragment.cartTotal = cartTotal
            return fragment
        }
    }

    fun setListener(listener: OnVoucherSelectedListener) {
        this.listener = listener
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_voucher_bottom_sheet, container, false)
        rvVouchers = view.findViewById(R.id.rvVouchers)
        etManualCode = view.findViewById(R.id.etManualCode)
        btnApplyManual = view.findViewById(R.id.btnApplyManual)

        rvVouchers.layoutManager = LinearLayoutManager(requireContext())

        btnApplyManual.setOnClickListener {
            val code = etManualCode.text.toString().trim()
            if (code.isNotEmpty() && listener != null) {
                listener?.onManualVoucherApplied(code.uppercase(Locale.getDefault()))
                dismiss()
            } else {
                Toast.makeText(context, "Vui lòng nhập mã hợp lệ", Toast.LENGTH_SHORT).show()
            }
        }

        setupAdapter()

        return view
    }

    private fun setupAdapter() {
        val available = ArrayList<VoucherItem>()
        val disabled = ArrayList<VoucherItem>()

        for (v in allVouchers) {
            if ("AVAILABLE" != v.status) {
                disabled.add(VoucherItem(v, 0.0, "Chỉ áp dụng với mã khả dụng / Còn hạn."))
                continue
            }
            if (v.minOrder > 0 && cartTotal < v.minOrder) {
                disabled.add(
                    VoucherItem(
                        v,
                        0.0,
                        "Đơn tối thiểu " + String.format(Locale.getDefault(), "%,.0fđ", v.minOrder)
                    )
                )
                continue
            }

            var actualDiscount = 0.0
            if ("PERCENTAGE" == v.discountType) {
                actualDiscount = (cartTotal * v.discountValue) / 100.0
            } else {
                actualDiscount = v.discountValue
            }
            if (v.maxDiscount > 0 && actualDiscount > v.maxDiscount) {
                actualDiscount = v.maxDiscount
            }
            available.add(VoucherItem(v, actualDiscount, null))
        }

        Collections.sort(available) { a, b -> b.actualDiscount.compareTo(a.actualDiscount) }

        val combined = ArrayList<VoucherItem>()
        combined.addAll(available)
        combined.addAll(disabled)

        rvVouchers.adapter = SheetAdapter(combined)
    }

    private inner class VoucherItem(
        val summary: VoucherSummary,
        val actualDiscount: Double,
        val disabledReason: String?
    )

    private inner class SheetAdapter(private val items: List<VoucherItem>) :
        RecyclerView.Adapter<SheetAdapter.VH>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
            val v = LayoutInflater.from(parent.context).inflate(R.layout.item_voucher, parent, false)
            return VH(v)
        }

        override fun onBindViewHolder(holder: VH, position: Int) {
            val item = items[position]
            val v = item.summary

            holder.tvCode.text = v.code
            holder.tvDescription.text = v.description
            holder.tvValidTo.text = "HSD: ${if (v.endDate != null) v.endDate.substring(0, 10) else "N/A"}"

            if (item.disabledReason == null) {
                // Khả dụng
                holder.itemView.alpha = 1.0f
                holder.tvStatus.text = "Khả dụng"
                holder.tvStatus.backgroundTintList = ColorStateList.valueOf(0x334CAF50)
                holder.tvStatus.setTextColor(-0xb350b0) // 0xFF4CAF50
                holder.tvDiscount.text = String.format(Locale.getDefault(), "-%,.0fđ", item.actualDiscount)

                holder.itemView.setOnClickListener {
                    listener?.onVoucherSelected(v)
                    dismiss()
                }
            } else {
                // Disabled
                holder.itemView.alpha = 0.5f
                holder.tvStatus.text = item.disabledReason
                holder.tvStatus.backgroundTintList = ColorStateList.valueOf(0x33888888.toInt())
                holder.tvStatus.setTextColor(0xFF888888.toInt())

                val discount = if ("PERCENTAGE".equals(v.discountType, ignoreCase = true)) {
                    "-${v.discountValue.toInt()}%"
                } else {
                    String.format(Locale.getDefault(), "-%,.0fđ", v.discountValue)
                }
                holder.tvDiscount.text = discount
                holder.itemView.setOnClickListener(null)
            }
        }

        override fun getItemCount(): Int = items.size

        inner class VH(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val tvCode: TextView = itemView.findViewById(R.id.tvCode)
            val tvDescription: TextView = itemView.findViewById(R.id.tvDescription)
            val tvValidTo: TextView = itemView.findViewById(R.id.tvValidTo)
            val tvDiscount: TextView = itemView.findViewById(R.id.tvDiscount)
            val tvStatus: TextView = itemView.findViewById(R.id.tvStatus)
        }
    }
}
